text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,113 @@
+- Start Date: 2023-02-01
+- RFC Type: feature
+- RFC PR: n/a
+- RFC Status: draft
+
+# Summary
+
+This RFC proposes introducing a centralized schema repository for the Kafka topics in use at Sentry. It provides mappings between the Kafka topics we are using to the encoding format and schemas for messages on that topic.
+
+Topic registration and schema validation of messages will be optional for the foreseeable future.
+
+# Motivation
+
+Kafka is increasingly used as the message bus between services at Sentry. Kafka topics and their schemas are usually not internal to any one service, but part of the contract between services.
+
+As Sentry grows, the number of message types and topics in use is exploding. We are marching further towards increasingly distributed ownership of data by product and infrastructure engineering teams. We have already seen an increase in the number of incidents related with invalid data and schema issues at Sentry. It is reasonable to expect this trend to continue as the number of topics, data types, teams and engineers contributing to Sentry grows.
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
```

Comment:
Two questions about backwards compatibility:
- I think we should define it a bit more formally. Maybe not everybody has the same concept of what can and cannot be done in mind.
- Do you plan to stick with one model of backward compatibility for the entire registry or is the compatibility model a property of a topic/development phase?

For the second question, for the sake of keeping development lean and quick, I think we should not force strict backward compatibility rule early on when a schema is being defined. 
Chances are, at this stage, that nobody is using it yet and more fundamental changes are required before a product starts actually relying on it. Think about the breaking changes we do on Clickhouse schemas before activating the migrations in CI and self hosted. 
I think we should have a way to tweak the schema easily early on and locking it down to strict backward compatibility rules as soon as we consider it viable.
Maybe a `dev` phase for each schema before it graduates to `stable` ?

Reply:
I think the idea of having various compatibility rules is a good one, as it allows for different types of development styles that may be appropriate at different products or times in product life cycle.

I will update the proposal to add a `compatibility_mode` key to the schema, and a couple of modes to start (perhaps "any" which will allow any change and "backward" or something like that).

The idea is that we could add additional modes in the future, if we want to, say, start enforcing stricter or different rules on specific topics

Reply:
ok